agent/grpc/stats_test.go

```diff
@@ -64,23 +65,42 @@ func TestHandler_EmitsStats(t *testing.T) {
 	// Wait for the server to stop so that active_streams is predictable.
 	require.NoError(t, g.Wait())
 
+	// Occasionally the active_stream=0 metric may be emitted before the
+	// active_conns=0 metric. The order of those metrics is not really important
+	// so we sort the calls to match the expected.
+	sort.Slice(sink.gaugeCalls, func(i, j int) bool {
+		if i < 2 || j < 2 {
+			return i < j
+		}
+		if len(sink.gaugeCalls[i].key) < 4 || len(sink.gaugeCalls[j].key) < 4 {
+			return i < j
+		}
+		return sink.gaugeCalls[i].key[3] < sink.gaugeCalls[j].key[3]
+	})
+
+	cmpMetricCalls := cmp.AllowUnexported(metricCall{})
 	expectedGauge := []metricCall{
 		{key: []string{"testing", "grpc", "server", "active_conns"}, val: 1},
 		{key: []string{"testing", "grpc", "server", "active_streams"}, val: 1},
 		{key: []string{"testing", "grpc", "server", "active_conns"}, val: 0},
 		{key: []string{"testing", "grpc", "server", "active_streams"}, val: 0},
 	}
-	require.Equal(t, expectedGauge, sink.gaugeCalls)
+	assertDeepEqual(t, expectedGauge, sink.gaugeCalls, cmpMetricCalls)
 
 	expectedCounter := []metricCall{
 		{key: []string{"testing", "grpc", "server", "request"}, val: 1},
 	}
-	require.Equal(t, expectedCounter, sink.incrCounterCalls)
+	assertDeepEqual(t, expectedCounter, sink.incrCounterCalls, cmpMetricCalls)
 }
 
-var fastRetry = &retry.Timer{Timeout: 7 * time.Second, Wait: 2 * time.Millisecond}
+func assertDeepEqual(t *testing.T, x, y interface{}, opts ...cmp.Option) {
+	t.Helper()
+	if diff := cmp.Diff(x, y, opts...); diff != "" {
+		t.Fatalf("assertion failed: values are not equal\n--- expected\n+++ actual\n%v", diff)
+	}
+}
 
-func patchGlobalMetrics(t *testing.T) *fakeMetricsSink {
+func patchGlobalMetrics(t *testing.T) (*fakeMetricsSink, func()) {
```

Comment:
I tried having this return a `reset` function instead of using `t.Cleanup` so that I could reset immediately after the `statsHandler` is created, but it did not really fix the flake.